### PR TITLE
WindowInsets

### LINKED ISSUE #9332: FLAG_FULLSCREEN deprecation in API 30
* FLAG_FULLSCREEN
  * Use WindowInsetsController.hide(int) with WindowInsets.Type.statusBars() instead.
  * Only the usage in `AnkiActivity` needs to be fixed. The usage in `Reviewer.kt` doesn't need to be fixed because the screen is deprecated 

---
The deprecations below don't need to be fixed because the screen that uses it is deprecated and will be eventually removed.

* ~~View.setOnSystemUiVisibilityChangeListener~~
   * ~~Use WindowInsets.isVisible(int) to find out about system bar visibilities by setting a View.OnApplyWindowInsetsListener on this view~~
* ~~View.getSystemUiVisibility()~~
  * ~~SystemUiVisibility flags are deprecated. Use WindowInsetsController instead.~~

### PR DESCRIPTION
**Pull request by n1snt**
## Purpose / Description
Deprecated UI flags were replaced by WindowInsets.

## Fixes
Fixes #9332

## How Has This Been Tested?
Ran tests & tested on:
- Physical Device (Android 12)
- Virtual Device (Android 7)

## Learning (optional, can help others)
I read [this ](https://medium.com/swlh/modifying-system-ui-visibility-in-android-11-e66a4128898b) guide & went through [this ](https://github.com/ankidroid/Anki-Android/pull/9406) pull request.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented on your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt [hunk 1]
```diff
@@ -37,6 +37,7 @@ class CompatHelper private constructor() {
         sdkVersion >= Build.VERSION_CODES.Q -> CompatV29()
         sdkVersion >= Build.VERSION_CODES.O -> CompatV26()
         sdkVersion >= Build.VERSION_CODES.M -> CompatV23()
+        sdkVersion >= Build.VERSION_CODES.R -> CompatV30()
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt:40]:
```suggestion
        sdkVersion >= Build.VERSION_CODES.R -> CompatV30()
        sdkVersion >= Build.VERSION_CODES.Q -> CompatV29()
        sdkVersion >= Build.VERSION_CODES.O -> CompatV26()
        sdkVersion >= Build.VERSION_CODES.M -> CompatV23()
```

These must be in decreasing order or the fall through doesn't work correctly

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package com.ichi2.compat
+
+import android.annotation.TargetApi
+import android.view.Window
+import android.view.WindowInsets
+import android.view.WindowInsetsController
+
+/** Implementation of [Compat] for SDK level 30 and higher. Check [Compat]'s for more detail.  */
+@TargetApi(30)
+class CompatV30 : CompatV26(), Compat {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.kt:25]:
```suggestion
class CompatV30 : CompatV29(), Compat {
```

must extend the next lower-one for the overriding to chain correctly and include the 29 overrides (and 28, and 27) vs 26

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt [hunk 3]
```diff
@@ -1286,28 +1288,49 @@ open class Reviewer : AbstractFlashcardViewer() {
...
-            // Note that system bars will only be "visible" if none of the
-            // LOW_PROFILE, HIDE_NAVIGATION, or FULLSCREEN flags are set.
-            val visible = flags and View.SYSTEM_UI_FLAG_HIDE_NAVIGATION == 0
-            Timber.d("System UI visibility change. Visible: %b", visible)
-            if (visible) {
-                showViewWithAnimation(toolbar)
-                if (fullscreenMode == FullScreenMode.FULLSCREEN_ALL_GONE) {
-                    showViewWithAnimation(topbar)
-                    showViewWithAnimation(answerButtons)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt:1291]: I don't understand why you have implemented the Compat base implementation (21) and "new style" (30) but then have a VERSION_CODES check here? It seems like you implemented the Compat stuff mostly correctly (the version chaining was off - but that's a quick fix) but then did not use it?

use the Compat stuff here and implement it in such a way that you do not need a version check - you just call the methods on the Compat object returned from CompatHelper and then this should shrink down a lot, with all the complexity in CompatV21/CompatV30

**Review by mikehardy — CHANGES_REQUESTED** [review]
Hey - this looks like a good start but not exactly the way Compat works. Probably  not too hard to reshape until it does work right though, it appears you have done the hard part of figuring out how the new API30 insets are supposed to work already and that is most of the task I think

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt (hunk 1 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt]: Done.

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.kt (hunk 2 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.kt]: Done.

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt (hunk 3 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt]: I have removed the VERSION_CODES checks.
But I'm not sure if I have done it correctly.
Can you please check this.

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt [hunk 4]
```diff
@@ -1270,54 +1270,19 @@ open class Reviewer : AbstractFlashcardViewer() {
...
-                    showViewWithAnimation(topbar)
-                    showViewWithAnimation(answerButtons)
-                }
-            } else {
-                hideViewWithAnimation(toolbar)
-                if (fullscreenMode == FullScreenMode.FULLSCREEN_ALL_GONE) {
-                    hideViewWithAnimation(topbar)
-                    hideViewWithAnimation(answerButtons)
-                }
-            }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt:1312]: This logic doesn't seem to map to the new compat code

**Review by david-allison — COMMENTED** [review]
Definite request; `CompatV31` needs to inherit from `CompatV30`.

I don't think the code is correct now for V21-V29.

Could a few videos be shown with the full screen behaviour?

**Comment by david-allison** [comment]
I know the feeling! no rush

**Comment by n1snt** [comment]
I forgot about this issue. Will fix the merge conflicts & start working on it again once I'm done with my exams.

**Pull request closed without merging**